extendr-api/src/iter.rs

```diff
@@ -89,14 +85,26 @@ impl Iterator for StrIter {
             let vector = self.vector.get();
             if i >= self.len {
                 None
-            } else if TYPEOF(vector) == SEXPTYPE::STRSXP {
-                Some(str_from_strsxp(vector, i as isize))
-            } else if TYPEOF(vector) == SEXPTYPE::INTSXP && TYPEOF(self.levels) == SEXPTYPE::STRSXP
-            {
-                let j = *(INTEGER(vector).add(i));
-                Some(str_from_strsxp(self.levels, j as isize - 1))
             } else if TYPEOF(vector) == SEXPTYPE::NILSXP {
                 Some(<&str>::na())
+            } else if TYPEOF(vector) == SEXPTYPE::STRSXP {
+                Some(str_from_strsxp(vector, i))
+            } else if TYPEOF(vector) == SEXPTYPE::CHARSXP {
+                if vector == R_NaString {
+                    Some(<&str>::na())
+                } else if vector == R_BlankString {
+                    Some("")
+                } else {
+                    Some(rstr::charsxp_to_str(vector))
+                }
+            } else if Rf_isFactor(vector).into() {
+                // factor support: factor is an integer, and we need
+                // the value of it, to retrieve the assigned label
+                let j = *(INTEGER(vector).add(i));
```

Comment:
Can't we do some sane transmute into `&[i32]` and then use indexer? This aint no C language! (I know it is written like this originally, but we can do better). This will probably cost us a bounds check at runtime, but I believe it is worth it for the readability sake. What do you think?

Reply:
You're absolutely right. I'm looking at it right now. There is an INTEGER_ELT... like bro, we should maybe use that.